File: scanner/iosApp/KvantScan/TaskListViewModel.swift
import Foundation
import shared

/// Список заданий поверх локального TaskStore — двойник Android-версии
/// (`androidApp/.../TaskListViewModel.kt`).
///
/// Офлайн-первичность (§04.1): список всегда читается из TaskStore; pull
/// лишь обновляет его. Ошибка сети — не пустой экран, а прежний список
/// с плашкой «не обновилось». Сортировка «срочные сверху» живёт в ядре.
@MainActor
final class TaskListViewModel: ObservableObject {

    @Published private(set) var tasks: [LocalTask] = []
    @Published private(set) var isRefreshing = false
    @Published private(set) var syncError: String?
    /// Всего записей в офлайн-каналах (события+медиа+трассировка).
    @Published private(set) var pendingSync: Int = 0

    private let store: TaskStore
    private let sync: SyncEngine
    /// Полный граф — для сводки «что ещё не уехало»; nil в лёгких тестах.
    private let stack: SyncStack?
    // Курсор пока в памяти процесса; персистентность придёт с SQLDelight.
    private var cursor: String?

    init(store: TaskStore, sync: SyncEngine, stack: SyncStack? = nil) {
        self.store = store
        self.sync = sync
        self.stack = stack
        Task {
            await reload()
            await refresh()
        }
    }

    func refresh() async {
        isRefreshing = true
        do {
            cursor = try await sync.pullOnce(
                cursor: cursor, applier: PullApplier(tasks: store))
            syncError = nil
        } catch {
            // Сеть упала — список остаётся прежним, это норма склада.
            syncError = (error as NSError).localizedDescription
        }
        await reload()
        isRefreshing = false
    }

    private func reload() async {
        if let current = try? await store.tasks() {
            tasks = current
        }
        if let stack, let status = try? await stack.syncStatus() {
            pendingSync = Int(status.total)
        }
    }
}

File: scanner/androidApp/src/main/kotlin/ru/kvant/scan/android/ui/TaskListViewModel.kt
package ru.kvant.scan.android.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import ru.kvant.scan.sync.LocalTask
import ru.kvant.scan.sync.PullApplier
import ru.kvant.scan.sync.SyncEngine
import ru.kvant.scan.sync.TaskStore
import ru.kvant.scan.sync.TransportException

/**
 * Список заданий поверх локального [TaskStore] (двойник — iOS
 * TaskListViewModel.swift). Сортировка «срочные сверху» живёт в ядре и
 * покрыта тестами; здесь — перекладка в StateFlow и запуск pull.
 *
 * Офлайн-первичность (§04.1): список всегда читается из TaskStore; pull
 * лишь обновляет его. Ошибка сети не пустой экран, а вчерашний список
 * с плашкой «не обновилось».
 */
class TaskListViewModel(
    private val tasks: TaskStore,
    private val sync: SyncEngine,
) : ViewModel() {

    data class UiState(
        val tasks: List<LocalTask> = emptyList(),
        val isRefreshing: Boolean = false,
        val syncError: String? = null,
    )

    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state.asStateFlow()

    // Курсор пока живёт в памяти процесса; персистентность — в directory_cache
    // по образцу справочников, когда подключится SQLDelight-драйвер.
    private var cursor: String? = null

    init {
        viewModelScope.launch { reload() }
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isRefreshing = true)
            try {
                cursor = sync.pullOnce(cursor, PullApplier(tasks))
                _state.value = _state.value.copy(syncError = null)
            } catch (e: TransportException) {
                // Сеть упала — список остаётся прежним, это норма склада.
                _state.value = _state.value.copy(syncError = e.message)
            }
            reload()
            _state.value = _state.value.copy(isRefreshing = false)
        }
    }

    private suspend fun reload() {
        _state.value = _state.value.copy(tasks = tasks.tasks())
    }
}
